src/typ/environment.rs

```diff
@@ -467,21 +482,113 @@ impl<'a, 'b> Environment<'a, 'b> {
-            self.warnings.push(Warning::UnusedType { name, location })
+        let unused = self.value_usages.pop_front().unwrap();
+        self.handle_unused_local_values(unused);
+    }
+
+    fn handle_unused_local_values(&mut self, unused: HashMap<String, (ValueKind, SrcSpan, usize)>) {
+        for val in unused {
+            match val {
+                (name, (ValueKind::ImportedType, location, 0)) => {
+                    self.warnings.push(Warning::UnusedType {
+                        name,
+                        imported: true,
+                        location,
+                    })
+                }
+                (name, (ValueKind::ImportedConstructor, location, 0)) => {
+                    self.warnings.push(Warning::UnusedConstructor {
+                        name,
+                        imported: true,
+                        location,
+                    })
+                }
+                (name, (ValueKind::ImportedTypeAndConstructor, location, 0)) => {
+                    self.warnings.push(Warning::UnusedType {
+                        name,
+                        imported: true,
+                        location,
+                    })
+                }
+                (name, (ValueKind::PrivateConstant, location, 0)) => self
+                    .warnings
+                    .push(Warning::UnusedPrivateModuleConstant { name, location }),
+                (name, (ValueKind::PrivateConstructor(_), location, 0)) => {
+                    self.warnings.push(Warning::UnusedConstructor {
+                        name,
+                        imported: false,
+                        location,
+                    })
+                }
+                (name, (ValueKind::PrivateFunction, location, 0)) => self
+                    .warnings
+                    .push(Warning::UnusedPrivateFunction { name, location }),
+                (name, (ValueKind::PrivateType, location, 0)) => {
+                    self.warnings.push(Warning::UnusedType {
+                        name,
+                        imported: false,
+                        location,
+                    })
+                }
+                (name, (ValueKind::ImportedValue, location, 0)) => self
+                    .warnings
+                    .push(Warning::UnusedImportedValue { name, location }),
+
+                (name, (ValueKind::Variable, location, 0)) => self
+                    .warnings
+                    .push(Warning::UnusedVariable { name, location }),
+
+                _ => {}
+            };
         }
-        for (name, location) in self.unused_private_mixed_constructors.drain() {
-            self.warnings
-                .push(Warning::UnusedConstructor { name, location })
+    }
+
+    pub fn init_value_usage(&mut self, name: String, kind: ValueKind, location: SrcSpan) {
+        match self
+            .value_usages
+            .front_mut()
+            .unwrap()
```

Comment:
Let's call `.gleam_expect` here do there's a nice message if it crashes due to a fatal bug.